cpp/community_detection_module/algorithm/label_propagation.cpp

```diff
@@ -0,0 +1,282 @@
+#include <cmath>
+#include <mg_graph.hpp>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "label_propagation.hpp"
+
+namespace LabelRankT {
+
+///@brief Checks if vector is subset of another vector.
+///
+///@tparam T -- set element data type
+///
+///@param v1 -- subset candidate
+///@param v2 -- vector against which the check is performed
+///
+///@return -- whether v1 is subset of v2
+template <class T>
+bool is_subset(std::vector<T> const& v1, std::vector<T> const& v2) {
+  for (const auto& element : v1) {
+    if (std::find(v2.begin(), v2.end(), element) == v2.end()) return false;
+  }
+
+  return true;
+}
+
+double LabelRankT::get_weight(std::uint64_t node_i_id,
+                              std::uint64_t node_j_id) {
+  // WORK IN PROGRESS
+
+  return DEFAULT_WEIGHT;
+}
+
+void LabelRankT::set_structures(std::uint64_t node_id) {
+  times_updated[node_id] = 0;
+  std::unordered_map<std::uint64_t, double> node_label_Ps;
+
+  double sum_w_i = w_selfloop;
+  for (const auto node_j : graph->Neighbours(node_id)) {
+    sum_w_i += get_weight(node_j.node_id, node_id);
+  }
+
+  // add self-loop
+  node_label_Ps[node_id] = w_selfloop / sum_w_i;
+
+  // add other edges
+  for (const auto node_j : graph->Neighbours(node_id)) {
+    node_label_Ps[node_j.node_id] =
+        get_weight(node_id, node_j.node_id) / sum_w_i;
+  }
+
+  label_Ps[node_id] = node_label_Ps;
+  sum_w[node_id] = sum_w_i;
+}
+
+void LabelRankT::remove_deleted_nodes(
+    std::unordered_set<std::uint64_t>& deleted_nodes_ids) {
+  for (const auto node_id : deleted_nodes_ids) {
+    label_Ps.erase(node_id);
+    sum_w.erase(node_id);
+    times_updated.erase(node_id);
+  }
+}
+
+void LabelRankT::reset_times_updated() {
+  for (auto [node_id, _] : times_updated) {
+    times_updated[node_id] = 0;
+  }
+}
+
+std::uint64_t LabelRankT::get_label(std::uint64_t node_id) {
+  std::uint64_t node_label;
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P) {
+      max_P = P;
+      node_label = label;
+    }
+  }
+
+  return node_label;
+}
+
+std::vector<std::uint64_t> LabelRankT::most_probable_labels(
+    std::uint64_t node_id) {
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P) max_P = P;
+  }
+
+  std::vector<std::uint64_t> most_probable_labels = {};
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P == max_P) most_probable_labels.push_back(label);
+  }
+
+  return most_probable_labels;
+}
+
+bool LabelRankT::distinct_enough(mg_graph::Node<uint64_t> node_i,
+                                 double similarity_threshold) {
+  std::vector<std::uint64_t> labels_i = most_probable_labels(node_i.id);
+  std::uint64_t label_similarity = 0;
+
+  int node_i_degree = 0;
+  for (const auto node_j : graph->Neighbours(node_i.id)) {
```

Comment:
Take a note here and think about the dynamic case. The neighbours in this case will be in the state after processing the query, meaning you wouldn't have deleted edges, and would have newly added ones. Hope that is fine for the algorithm

Reply:
This is in line with the algorithm design :+1: